File: app/src/main/java/com/zy/ppmusic/mvp/model/MediaActivityModelImpl.java
package com.zy.ppmusic.mvp.model;

import android.content.Context;
import android.os.Bundle;
import android.os.ResultReceiver;
import android.support.annotation.NonNull;
import android.support.v4.media.session.MediaControllerCompat;

import com.zy.ppmusic.mvp.contract.IMediaActivityContract;
import com.zy.ppmusic.utils.FileUtils;
import com.zy.ppmusic.utils.ScanMusicFile;

import java.lang.ref.WeakReference;
import java.util.concurrent.LinkedBlockingDeque;
import java.util.concurrent.ThreadFactory;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

/**
 * @author ZhiTouPC
 */
public class MediaActivityModelImpl implements IMediaActivityContract.IMediaActivityModel {
    private static final String TAG = "MediaMediaActivityModel";
    private ThreadPoolExecutor mBackGroundPool;

    public MediaActivityModelImpl() {
        mBackGroundPool = new ThreadPoolExecutor(1, 4, 0L,
                TimeUnit.MICROSECONDS, new LinkedBlockingDeque<Runnable>(), new ThreadFactory() {
            @Override
            public Thread newThread(@NonNull Runnable r) {
                return new Thread(r, TAG);
            }
        });
    }

    @Override
    public void refreshQueue(Context context, ScanMusicFile.AbstractOnScanComplete complete) {
        ScanMusicFile.getInstance().setOnScanComplete(complete).scanMusicFile(context);
    }

    @Override
    public void loadLocalData(String path, final IOnLocalDataLoadFinished callback) {
        final String localPath = path;
        mBackGroundPool.execute(new Runnable() {
            @Override
            public void run() {
                Object localData = FileUtils.readObject(localPath);
                callback.callBack(localData);
            }
        });
    }

    @Override
    public void postSendCommand(MediaControllerCompat controller, final String method, final Bundle params,
                                ResultReceiver resultReceiver) {
        final WeakReference<MediaControllerCompat> mWeakController = new WeakReference<>(controller);
        final WeakReference<ResultReceiver> mWeakReceiver = new WeakReference<>(resultReceiver);
        mBackGroundPool.execute(new Runnable() {
            @Override
            public void run() {
                if(mWeakController.get() != null){
                    mWeakController.get().sendCommand(method,params,mWeakReceiver.get());
                }
            }
        });
    }

    @Override
    public void postPlayWithId(MediaControllerCompat controller, final String mediaId, final Bundle extra) {
        final WeakReference<MediaControllerCompat> mWeakController = new WeakReference<>(controller);
        mBackGroundPool.execute(new Runnable() {
            @Override
            public void run() {
                if(mWeakController.get() != null){
                    mWeakController.get().getTransportControls().playFromMediaId(mediaId,extra);
                }
                mWeakController.clear();
            }
        });
    }

    @Override
    public void postSendCustomAction(MediaControllerCompat controller, final String action, final Bundle extra) {
        final WeakReference<MediaControllerCompat> mWeakController = new WeakReference<>(controller);
//        mBackGroundPool.execute(new Runnable() {
//            @Override
//            public void run() {
                if(mWeakController.get() != null){
                    mWeakController.get().getTransportControls().sendCustomAction(action,extra);
                }
                mWeakController.clear();
//            }
//        });
    }

    @Override
    public void postSetRepeatMode(MediaControllerCompat controller, final int mode) {
        final WeakReference<MediaControllerCompat> mWeakController = new WeakReference<>(controller);
        mBackGroundPool.execute(new Runnable() {
            @Override
            public void run() {
                if(mWeakController.get() != null){
                    mWeakController.get().getTransportControls().setRepeatMode(mode);
                }
                mWeakController.clear();
            }
        });
    }

    @Override
    public void postSkipNext(MediaControllerCompat controller) {
        final WeakReference<MediaControllerCompat> mWeakController = new WeakReference<>(controller);
        mBackGroundPool.execute(new Runnable() {
            @Override
            public void run() {
                if(mWeakController.get() != null){
                    mWeakController.get().getTransportControls().skipToNext();
                }
                mWeakController.clear();
            }
        });
    }

    @Override
    public void postSkipPrevious(MediaControllerCompat controller) {
        final WeakReference<MediaControllerCompat> mWeakController = new WeakReference<>(controller);
        mBackGroundPool.execute(new Runnable() {
            @Override
            public void run() {
                if(mWeakController.get() != null){
                    mWeakController.get().getTransportControls().skipToPrevious();
                }
                mWeakController.clear();
            }
        });
    }

    @Override
    public void postSkipToPosition(MediaControllerCompat controller, final Long id) {
        final WeakReference<MediaControllerCompat> mWeakController = new WeakReference<>(controller);
        mBackGroundPool.execute(new Runnable() {
            @Override
            public void run() {
                if(mWeakController.get() != null){
                    mWeakController.get().getTransportControls().skipToQueueItem(id);
                }
                mWeakController.clear();
            }
        });
    }

    @Override
    public void shutdown() {
        if (mBackGroundPool != null) {
            mBackGroundPool.shutdown();
            mBackGroundPool = null;
        }
    }

}

File: app/src/main/java/com/zy/ppmusic/mvp/model/MediaActivityModelImpl.kt
package com.zy.ppmusic.mvp.model

import android.content.Context
import android.content.SharedPreferences
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.os.ResultReceiver
import android.support.v4.media.session.MediaControllerCompat
import android.support.v4.media.session.PlaybackStateCompat
import com.zy.ppmusic.entity.MusicInfoEntity
import com.zy.ppmusic.mvp.contract.IMediaActivityContract
import com.zy.ppmusic.utils.*
import kotlinx.coroutines.*
import java.lang.ref.WeakReference

/**
 * @author stealfeam
 */
class MediaActivityModelImpl : IMediaActivityContract.IMediaActivityModel {
    private val CACHE_MODE_NAME = "cache_mode"
    private val CACHE_MODE_KEY = "mode_key"
    private val CACHE_ROOT_URI = "root_uri"
    private val CACHE_CHILD_URI = "child_uri"

    private val scope: CoroutineScope by lazy { CoroutineScope(Dispatchers.IO + Job()) }

    @Volatile
    private var mCachePreference: SharedPreferences? = null
    private var mControllerWeak: WeakReference<MediaControllerCompat>? = null

    private fun initCachePreference(context: Context) {
        if (mCachePreference == null) {
            mCachePreference = context.getSharedPreferences(CACHE_MODE_NAME, Context.MODE_PRIVATE)
        }
    }

    private fun changeLocalMode(c: Context, mode: Int) {
        initCachePreference(c)
        if (mCachePreference!!.getInt(CACHE_MODE_KEY, PlaybackStateCompat.REPEAT_MODE_NONE) == mode) {
            return
        }
        mCachePreference?.edit()?.putInt(CACHE_MODE_KEY, mode)?.apply()
    }

    override fun getGrantedRootUri(): String {
        return mCachePreference?.getString(CACHE_ROOT_URI, "") ?: ""
    }

    override fun getGrantedChildrenUri(): String {
        return mCachePreference?.getString(CACHE_CHILD_URI, "") ?: ""
    }

    override fun saveGrantedUri(root: String, child: String) {
        mCachePreference?.edit()?.apply {
            putString(CACHE_ROOT_URI, root)
            putString(CACHE_CHILD_URI, child)
        }?.apply()
    }

    override fun getLocalMode(c: Context, e: (Int) -> Unit) {
        //创建协程环境
        //CoroutineScope注释代码
        scope.launch(Dispatchers.Main) {
            val job = async(Dispatchers.IO){
                initCachePreference(c)
                return@async mCachePreference?.getInt(CACHE_MODE_KEY, PlaybackStateCompat.REPEAT_MODE_NONE) ?: 0
            }
            val mode = job.await()
            e.invoke(mode)
        }

    }

    override fun removeQueueItem(position: Int) {
        scope.launch(Dispatchers.IO) {
            mControllerWeak?.get()?.removeQueueItem(DataProvider.get().queueItemList.get()[position].description)
        }
    }

    override fun attachController(controller: MediaControllerCompat?) {
        controller?.apply {
            mControllerWeak = WeakReference(controller)
        }
    }

    override fun postSendCommand(method: String, params: Bundle,
                                 resultReceiver: ResultReceiver) {
        scope.launch(Dispatchers.IO) {
            mControllerWeak?.get()?.sendCommand(method, params, resultReceiver)
        }
    }

    override fun postPlayWithId(mediaId: String, extra: Bundle) {
        scope.launch(Dispatchers.IO) {
            mControllerWeak?.get()?.transportControls?.playFromMediaId(mediaId, extra)
        }
    }

    override fun postSendCustomAction(action: String, extra: Bundle) {
        mControllerWeak?.get()?.transportControls?.sendCustomAction(action, extra)
    }


    override fun postSetRepeatMode(c: Context, mode: Int) {
        scope.launch(Dispatchers.IO) {
            mControllerWeak?.get()?.transportControls?.setRepeatMode(mode)
            changeLocalMode(c, mode)
        }
    }

    override fun postSkipNext() {
        mControllerWeak?.get()?.transportControls?.let {
            scope.launch(Dispatchers.IO) {
                it.skipToNext()
            }
        }
    }

    override fun postSkipPrevious() {
        scope.launch(Dispatchers.IO) {
            mControllerWeak?.get()?.transportControls?.skipToPrevious()
        }
    }

    override fun postSkipToPosition(id: Long) {
        scope.launch(Dispatchers.IO) {
            PrintLog.e("Model准备传输-----")
            mControllerWeak?.get()?.transportControls?.skipToQueueItem(id)
        }
    }

    override fun shutdown() {
        scope.cancel()
        mControllerWeak?.clear()
    }

}
